music_grabber.py

```diff
@@ -0,0 +1,123 @@
+        ydl_opts = {
+                'format': 'bestaudio[protocol^=http]',
+                'outtmpl': DOWNLOAD_DIR + '%(extractor)s_%(id)s.%(ext)s',
+                'writethumbnail': True
+        }
+        ydl = YoutubeDL(ydl_opts)
+        info = ydl.extract_info(input_url, download=False)
+        webpage_info = info['extractor'] + " (" + info['id'] + ")"
+        info_format = next((item for item in info['formats']
+                            if item['format_id'] == info['format_id']),
+                           None)
+
+        if self.__youtubeVideoNotMusic(info):
+            raise WrongCategoryError(webpage_info +
+                                     ": not under Music category")
+
+        audio_filesize = self.__getFileSize(info_format)
+        if int(audio_filesize) < filesize_limit:
+            logger.info(webpage_info + ": downloading...")
+            ydl.process_info(info)
+            info = self.__addDownloadsSubset(info)
+            if self.__checkDownloadedFiles(info['downloads']):
+                return info
+            else:
+                raise ValueError(webpage_info +
+                                 ": failed to download audio or thumbnail")
+        else:
+            raise WrongFileSizeError(webpage_info + ": audio file size (" +
+                                     str(audio_filesize) + ") is greater" +
+                                     "than the limit: (" +
+                                     str(filesize_limit) + ")")
+
+    def __youtubeVideoNotMusic(self, info):
+        if info['extractor'] == 'youtube' and \
+                'Music' not in info['categories']:
+            return True
+        else:
+            return False
+
+    def __getFileSize(self, info_format):
+        if 'filesize' in info_format:
+            audio_filesize = info_format['filesize']
+        else:
+            audio_url = info_format['url']
+            http_headers = info_format['http_headers']
+            request_head = Request(audio_url, method='HEAD',
+                                   headers=http_headers)
+            audio_filesize = urlopen(request_head).headers['Content-Length']
+        return audio_filesize
+
+    def __addDownloadsSubset(self, info):
+        webpage_id = info['id']
+        extractor = info['extractor']
+        thumbnail_url = info['thumbnail']
+        basename = DOWNLOAD_DIR + extractor + '_' + webpage_id
+        thumbnail_file = basename + "." + \
+            self.__getFileExtensionFromUrl(thumbnail_url)
+        audio_file = basename + '.' + info['ext']
+        info['downloads'] = {
+                'audio': audio_file,
+                'thumbnail': thumbnail_file
+        }
+        return info
+
+    def __getFileExtensionFromUrl(self, url):
+        url_path = urlparse(url).path
+        basename = os.path.basename(url_path)
+        ext = basename.split(".")[-1]
+        return ext
+
+    def __checkDownloadedFiles(self, info_downloads):
```

Comment:
method called __**check**DownloadedFiles but it doesn't just check the files, it also deletes them. I guess you need to come up with a better name for a method

Reply:
done. changed to `__checkAndMayCleanDownloads`

Reply:
actually calling method as if it does two actions is also a bad practice, because ideally method should do one thing (the thing that its name suggests). This method is a bit confusing tbh, why do you need to delete downloads after downloading?

Reply:
from what i understand you check if you downloaded the files successfully and if one of the files is missing then you delete the other. That makes sense. So on this case you can do something like this
```
def __checkFilesSuccessfullyDownloaded(self, info_downloads):
         return all(os.path.isfile(f) for f in info_downloads.values())
```
and then in **download** method
```
            if self. __checkFilesSuccessfullyDownloaded(info['downloads']):
                 return info
             else:
                 __deleteDownloadedFiles(info['downloads'])
                 raise ValueError(webpage_info +
                                  ": failed to download audio or thumbnail")
```
in this case **__deleteDownloadedFiles** will contain this logic
```
for f in info_downloads.values():
                 try:
                     os.remove(f)
                 except OSError:
                     pass
```